src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.h

```diff
@@ -155,49 +219,79 @@ struct BallPlacementPlayFSM
 
         DEFINE_SML_ACTION(alignPlacement)
         DEFINE_SML_ACTION(placeBall)
-        DEFINE_SML_ACTION(kickOffWall)
+        DEFINE_SML_ACTION(setPickOffDest)
+        DEFINE_SML_ACTION(alignWall)
+        DEFINE_SML_ACTION(pickOffWall)
+        DEFINE_SML_ACTION(releaseBall)
         DEFINE_SML_ACTION(startWait)
         DEFINE_SML_ACTION(retreat)
+        DEFINE_SML_ACTION(wait)
 
-        DEFINE_SML_GUARD(shouldKickOffWall)
+        DEFINE_SML_GUARD(shouldPickOffWall)
         DEFINE_SML_GUARD(alignDone)
-        DEFINE_SML_GUARD(kickDone)
+        DEFINE_SML_GUARD(wallAlignDone)
+        DEFINE_SML_GUARD(wallPickOffDone)
+        DEFINE_SML_GUARD(ballReleased)
         DEFINE_SML_GUARD(ballPlaced)
         DEFINE_SML_GUARD(waitDone)
         DEFINE_SML_GUARD(retreatDone)
 
         return make_transition_table(
             // src_state + event [guard] / action = dest_state
-            *StartState_S + Update_E[!shouldKickOffWall_G] / alignPlacement_A =
+            *StartState_S + Update_E[!shouldPickOffWall_G] / alignPlacement_A =
                 AlignPlacementState_S,
-            StartState_S + Update_E[shouldKickOffWall_G] = KickOffWallState_S,
-            KickOffWallState_S + Update_E[!kickDone_G && shouldKickOffWall_G] /
-                                     kickOffWall_A                = KickOffWallState_S,
-            KickOffWallState_S + Update_E[kickDone_G]             = KickOffWallState_S,
-            KickOffWallState_S + Update_E[!kickDone_G]            = AlignPlacementState_S,
-            AlignPlacementState_S + Update_E[shouldKickOffWall_G] = KickOffWallState_S,
+            StartState_S + Update_E[shouldPickOffWall_G] = AlignWallState_S,
+
+            AlignWallState_S + Update_E[!wallAlignDone_G && shouldPickOffWall_G] /
+                                   alignWall_A = AlignWallState_S,
+            AlignWallState_S + Update_E[wallAlignDone_G] / setPickOffDest_A =
+                PickOffWallState_S,
+            AlignWallState_S + Update_E[!shouldPickOffWall_G] = AlignPlacementState_S,
+
+            PickOffWallState_S + Update_E[!wallPickOffDone_G] / pickOffWall_A =
+                PickOffWallState_S,
+            PickOffWallState_S + Update_E[wallPickOffDone_G] / startWait_A =
+                PickOffWaitState_S,
+            PickOffWaitState_S + Update_E[!waitDone_G] / wait_A = PickOffWaitState_S,
+            PickOffWaitState_S + Update_E[waitDone_G]           = ReleaseWallState_S,
```

Comment:
i don't think you need a separate `ReleaseWallState`, but instead reuse the `ReleaseState` I mentioned below. it can go something like this

1. At any point, if you're in Release state -> try to release the ball
2. Once release is done, 
3. If ball is at the placement location -> retreat
4. If not, move to `AlignPlacementState_S` and continue the placement